File: python/data-clumps/bad/customer_service.py
class CustomerService:
    def create_customer(self, first_name, last_name, email, street, city, state, zip_code, phone, date_of_birth):
        # Validate data
        if not first_name or not last_name:
            raise Exception('Name is required')
        if not self._is_valid_email(email):
            raise Exception('Invalid email')

        # Create customer record
        customer_data = {
            'first_name': first_name,
            'last_name': last_name,
            'email': email,
            'street': street,
            'city': city,
            'state': state,
            'zip_code': zip_code,
            'phone': phone,
            'date_of_birth': date_of_birth
        }

        # Save to database (simulated)
        return customer_data

    def update_customer_address(self, customer_id, street, city, state, zip_code):
        # Update address
        address_data = {
            'street': street,
            'city': city,
            'state': state,
            'zip_code': zip_code
        }

        # Save to database (simulated)
        return {'customer_id': customer_id, **address_data}

    def send_welcome_email(self, first_name, last_name, email, street, city, state, zip_code):
        full_name = f"{first_name} {last_name}"
        full_address = f"{street}, {city}, {state} {zip_code}"

        message = f"Welcome {full_name}!\n\n"
        message += f"Your address: {full_address}\n"

        # Send email (simulated)
        return {'to': email, 'message': message}

    def validate_shipping_address(self, street, city, state, zip_code):
        if not street or not city or not state or not zip_code:
            return False

        # Additional validation logic
        if len(zip_code) != 5:
            return False

        return True

    def format_address_label(self, first_name, last_name, street, city, state, zip_code):
        full_name = f"{first_name} {last_name}"
        full_address = f"{street}\n{city}, {state} {zip_code}"

        return f"{full_name}\n{full_address}"

    def _is_valid_email(self, email):
        import re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None

File: java/data-clumps/bad/CustomerService.java
import java.util.*;

public class CustomerService {
    public Map<String, Object> createCustomer(
            String firstName, String lastName, String email,
            String street, String city, String state, String zipCode,
            String phone, String dateOfBirth
    ) throws Exception {
        // Validate data
        if (firstName == null || firstName.isEmpty() || lastName == null || lastName.isEmpty()) {
            throw new Exception("Name is required");
        }
        if (!isValidEmail(email)) {
            throw new Exception("Invalid email");
        }

        // Create customer record
        Map<String, Object> customerData = new HashMap<>();
        customerData.put("first_name", firstName);
        customerData.put("last_name", lastName);
        customerData.put("email", email);
        customerData.put("street", street);
        customerData.put("city", city);
        customerData.put("state", state);
        customerData.put("zip_code", zipCode);
        customerData.put("phone", phone);
        customerData.put("date_of_birth", dateOfBirth);

        // Save to database (simulated)
        return customerData;
    }

    public Map<String, Object> updateCustomerAddress(
            int customerId,
            String street, String city, String state, String zipCode
    ) {
        // Update address
        Map<String, Object> addressData = new HashMap<>();
        addressData.put("street", street);
        addressData.put("city", city);
        addressData.put("state", state);
        addressData.put("zip_code", zipCode);

        // Save to database (simulated)
        Map<String, Object> result = new HashMap<>(addressData);
        result.put("customer_id", customerId);
        return result;
    }

    public Map<String, String> sendWelcomeEmail(
            String firstName, String lastName, String email,
            String street, String city, String state, String zipCode
    ) {
        String fullName = firstName + " " + lastName;
        String fullAddress = street + ", " + city + ", " + state + " " + zipCode;

        String message = "Welcome " + fullName + "!\n\n";
        message += "Your address: " + fullAddress + "\n";

        // Send email (simulated)
        Map<String, String> result = new HashMap<>();
        result.put("to", email);
        result.put("message", message);
        return result;
    }

    public boolean validateShippingAddress(
            String street, String city, String state, String zipCode
    ) {
        if (street == null || street.isEmpty() ||
            city == null || city.isEmpty() ||
            state == null || state.isEmpty() ||
            zipCode == null || zipCode.isEmpty()) {
            return false;
        }

        // Additional validation logic
        if (zipCode.length() != 5) {
            return false;
        }

        return true;
    }

    public String formatAddressLabel(
            String firstName, String lastName,
            String street, String city, String state, String zipCode
    ) {
        String fullName = firstName + " " + lastName;
        String fullAddress = street + "\n" + city + ", " + state + " " + zipCode;

        return fullName + "\n" + fullAddress;
    }

    private boolean isValidEmail(String email) {
        // Simple email validation
        return email != null && email.contains("@") && email.contains(".");
    }
}
